pkg/globalnet/controllers/ipam/gatewaymonitor.go

```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
+	endpoint := obj.(*v1.Endpoint)
 
-				return nil
-			}
+	klog.V(log.DEBUG).Infof("In processNextEndpoint, endpoint info: %+v", endpoint)
 
-			for _, remoteSubnet := range endpoint.Spec.Subnets {
-				MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-			}
+	if endpoint.Spec.ClusterID != gm.clusterID {
+		klog.V(log.DEBUG).Infof("Endpoint %s belongs to a remote cluster", endpoint.Spec.Hostname)
 
-			gm.endpointWorkqueue.Forget(obj)
-
-			return nil
-		}
-
-		endpoints, err := gm.submarinerClientSet.SubmarinerV1().Endpoints(ns).List(context.TODO(), metav1.ListOptions{})
+		overlap, err := cidr.IsOverlapping(endpoint.Spec.Subnets, gm.ipamSpec.GlobalCIDR[0])
 		if err != nil {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error retrieving submariner endpoint list %v", err)
+			// Ideally this case will never hit, as the subnets are valid CIDRs
+			klog.Warningf("unable to validate overlapping Service CIDR: %s", err)
 		}
 
-		for _, endpoint := range endpoints.Items {
-			if endpoint.Spec.ClusterID != gm.clusterID {
-				for _, remoteSubnet := range endpoint.Spec.Subnets {
-					MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-				}
-			}
-		}
+		if overlap {
+			// When GlobalNet is used, globalCIDRs allocated to the clusters should not overlap.
+			// If they overlap, skip the endpoint as its an invalid configuration which is not supported.
+			klog.Errorf("GlobalCIDR %q of local cluster %q overlaps with remote cluster %s",
+				gm.ipamSpec.GlobalCIDR[0], gm.ipamSpec.ClusterID, endpoint.Spec.ClusterID)
 
-		hostname, err := os.Hostname()
-		if err != nil {
-			klog.Fatalf("Unable to determine hostname: %v", err)
+			return false
 		}
 
-		// If the endpoint hostname matches with our hostname, it implies we are on gateway node
-		if endpoint.Spec.Hostname == hostname {
-			klog.V(log.DEBUG).Infof("We are now on GatewayNode %s", endpoint.Spec.PrivateIP)
-
-			configureTCPMTUProbe()
-
-			gm.syncMutex.Lock()
-			if !gm.isGatewayNode {
-				gm.isGatewayNode = true
-				gm.initializeIPAMController(gm.ipamSpec.GlobalCIDR[0], gm.nodeName)
-			}
-			gm.syncMutex.Unlock()
-		} else {
-			klog.V(log.DEBUG).Infof("We are on non-gatewayNode. GatewayNode ip is %s", endpoint.Spec.PrivateIP)
-
-			gm.syncMutex.Lock()
-			if gm.isGatewayNode {
-				gm.stopIPAMController()
-				gm.isGatewayNode = false
-			}
-			gm.syncMutex.Unlock()
+		for _, remoteSubnet := range endpoint.Spec.Subnets {
+			MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
 		}
 
-		gm.endpointWorkqueue.Forget(obj)
-
-		return nil
-	}()
+		return false
+	}
 
+	hostname, err := os.Hostname()
 	if err != nil {
-		utilruntime.HandleError(err)
-		return true
+		klog.Fatalf("Unable to determine hostname: %v", err)
 	}
 
-	return true
-}
+	// If the endpoint hostname matches with our hostname, it implies we are on gateway node
+	if endpoint.Spec.Hostname == hostname {
```

Comment:
This detection is also provided by the event interface. (Transition to gateway and transition to non gateway)